lib/index.js

```
@@ -1,5 +1,9 @@
 /* eslint no-process-env: 0, no-console: 0, max-statements: 0 */
 'use strict';
+process.on('unhandledRejection', function(e) {
+  // no swallowed Promise errors
+  throw e;
```

So what happens when this throws? Does that cause an `unhandledException` error to be raised? And if so, what would be the consequences of that?

What problem are we trying to solve here? Will this crash the gulp tasks? It seems like the responsibility of an application to decide how to deal with global unhandled errors, and maybe we wouldn't want whatever this causes in all cases.

---

> Does that cause an unhandledException error to be raised? And if so, what would be the consequences of that?

I would trigger unhandledException if there was a listener attached (which is highly likely in a mocha test). The consequences are normal exception handling instead of swallowed errors.

> What problem are we trying to solve here?

We are trying to prevent errors that are not directly from a Promise from being swallowed by the Promise handler, such as when one of the functions in an `async.waterfall` operation contains a Promise and a different function throws an error, the Promise handling would result in the error being swallowed by the promise because of the control flow of ES6 Promises.

> Will this crash the gulp tasks?

Hopefully

> It seems like the responsibility of an application to decide how to deal with global unhandled errors, and maybe we wouldn't want whatever this causes in all cases.

This specifically targets test tools that are designed to aid us in building useful testing scenarios, this does not discourage the appropriate handling of Promise rejections, it only reduces the friction between seemingly unrelated errors and Promises capturing those errors in the rejection handling and then not displaying them because of a lack of proper catch handling, which is more complex than it appears on the surface, because you have to place the handler in the appropriate place based on wear the rejection/exception occurs.


I would be happy to move this into the `runMocha` function if you prefer, but I don't think that it hurts anything to apply to all tasks.

---

I see what you mean. I wasn't aware people were using "swallows" for scenarios like this, makes sense:

```es6
somePromise.then(() => {
  someThingThatThrows();
}); // no catch!
```

I wouldn't say that the Promise is swallowing the error, only that the developer is ignoring the error.

:+1: if this can provide visibility to ignored errors during the running of tests, but I think it'd be better if it could be turned on with an option in case it causes issues.

---

@DullReferenceException I have a much stranger case too.
```javascript
function collaborate(done) {
async.waterfall([
  (cb) => {
    promiseThing.then(c => {
      return otherPromiseThing.then(r => cb(null, r)); // where the exception was being caught, or not caught
    }).catch(e => cb(e));
  },
  (r, cb) => {
    cb(null, syncThing(r));
  }
], (e, u) => {
  if (err) return void done(e);
  done(null, u); // where the exception occurs
});
}

---

Yeah, cases like that is where Bluebird's `.asCallback()` is useful for interop; this is subtly broken, and it's not apparent how. I'm glad you're adding something to help us potentially catch stuff like this.